openretina/dataloaders.py

```diff
@@ -50,76 +52,150 @@ def __len__(self):
+        return iter(np.array(shifted_indices)[indices_shuffling])
 
     def __len__(self):
         return len(self.indices)
 
 
+def gen_shifts(clip_bounds, start_indices, clip_chunk_size=50, allow_over_boundaries=False):
+    """
+    Generate shifted indices based on clip bounds and start indices.
+    Assumes that the original start indices are already within the clip bounds.
+    If they are not, it changes the overflowing indexes to respect the closest bound.
+
+    Args:
+        clip_bounds (list): A list of clip bounds.
+        start_indices (list): A list of start indices.
+        clip_chunk_size (int, optional): The size of each clip chunk. Defaults to 50.
+        allow_over_boundaries (bool, optional): Whether to allow training clips to be
+                                sampled over single-clip boundaries. Defaults to False.
+
+    Returns:
+        list: A list of shifted indices.
+
+    """
+
+    def get_next_bound(value, bounds):
+        insertion_index = bisect.bisect_right(bounds, value)
+        return bounds[min(insertion_index, len(bounds) - 1)]
+
+    shifted_indices = []
+    shifts = np.random.randint(1, clip_chunk_size // 2, len(start_indices))
+
+    for i, start_idx in enumerate(start_indices):
+        next_bound = get_next_bound(start_idx, clip_bounds)
+        if allow_over_boundaries:
+            shifted_indices.append(start_idx + shifts[i])
+        else:
+            if start_idx + shifts[i] + clip_chunk_size < next_bound:
+                shifted_indices.append(start_idx + shifts[i])
+            elif start_idx + clip_chunk_size > next_bound:
+                shifted_indices.append(next_bound - clip_chunk_size)
+            else:
+                shifted_indices.append(start_idx)
+
+    # Ensure we do not exceed the movie length when allowing over boundaries
+    if allow_over_boundaries and shifted_indices[-1] + clip_chunk_size > clip_bounds[-1]:
+        shifted_indices[-1] = clip_bounds[-1] - clip_chunk_size
+    return shifted_indices
+
+
 def get_movie_dataloader(
     movies: np.ndarray | torch.Tensor | dict[int, np.ndarray],
-    responses: Float[np.ndarray, "n_frames n_neurons"],  # noqa
-    roi_ids: Optional[Float[np.ndarray, "n_neurons"]],  # noqa
-    roi_coords: Optional[Float[np.ndarray, "n_neurons 2"]],  # noqa
-    group_assignment: Optional[Float[np.ndarray, "n_neurons"]],  # noqa
+    responses: Float[np.ndarray, "n_frames n_neurons"],
+    roi_ids: Optional[Float[np.ndarray, " n_neurons"]],
+    roi_coords: Optional[Float[np.ndarray, "n_neurons 2"]],
+    group_assignment: Optional[Float[np.ndarray, " n_neurons"]],
     split: str,
     start_indices: List[int] | Dict[int, List[int]],
     scan_sequence_idx: Optional[int] = None,
     chunk_size: int = 50,
     batch_size: int = 32,
     scene_length: Optional[int] = None,
     drop_last=True,
+    use_base_sequence=False,
+    allow_over_boundaries=True,
```

Comment:
```suggestion
    drop_last: bool = True,
    use_base_sequence: bool = False,
    allow_over_boundaries: bool = True,
```

maybe add type here, too